juju/juju
# log forwarding: Workers now run per-model

## Description of change

At the moment, the most expensive step in destroying a model is clearing out its logs. We want to make this cheaper by having a log collection for each model. But in order to do this we have to make log forwarding happen per-model rather than running for all models at once.

Move the log forwarder worker creation into the model manifolds, and allow the log streaming endpoint to be called for non-controller models. Remove the all-model capability of the log tailer, and always include the model UUID in the log messages it generates.

## QA steps

* Generate a CA cert + key and certs and keys for a server and client. 
* Set up an rsyslog server to receive forwarded messages with the server certificate. 
  * Refer to [this tutorial](http://www.rsyslog.com/doc/v8-stable/tutorials/tls_cert_summary.html) for information on this. 
  * Be sure to match the common names in the certificates to the names you put into the rsyslog config.
  * If you're specifying the IP address of the server, ensure the server certificate has the right IP address as a SubjectAltName - Go TLS support enforces this although OpenSSL doesn't by default.
* Bootstrap a controller and deploy some applications and relations to the default model.
* Configure log forwarding by specifying the following model-config for both controller and default model:
  * `syslog-host`: the IP and port of the rsyslog receiver (10514 if you follow the instructions above).
  * `syslog-ca-cert`
  * `syslog-client-cert`
  * `syslog-client-key`
* Turn forwarding on for only the controller by setting `logforward-enable` to true.
* Check that log messages from the controller appear in the receiver's logs, and that messages from the default model don't.
* Turn off logging for the controller model and enable it for the default model - records from the default model should start coming in and the controller ones should stop.
